### PR TITLE
Avoid some type instability issues

### PR DESCRIPTION
**Pull request by fingolfin**
... resp. trigger points for method invalidations

### CODE REVIEW

File: src/AlgAss/AlgMat.jl [hunk 1]
```diff
@@ -12,9 +12,9 @@ base_ring(A::MatAlgebra{T, S}) where {T, S} = A.base_ring::parent_type(T)
 
 base_ring_type(::Type{MatAlgebra{T, S}}) where {T, S} = parent_type(T)
 
-coefficient_ring(A::MatAlgebra) = A.coefficient_ring
+coefficient_ring(A::MatAlgebra{T, <:MatrixElem{S}}) where {T, S} = A.coefficient_ring::base_ring_type(S)
```
Comment (lgoettgens) [inline src/AlgAss/AlgMat.jl:15]:
```suggestion
coefficient_ring(A::MatAlgebra{T, <:MatrixElem{S}}) where {T, S} = A.coefficient_ring::parent_type(S)
```
Or alternatively take the base_ring_type of the second type param of MatAlgebra

Comment (fingolfin) [reply src/AlgAss/AlgMat.jl:15]: Ah, well spotted, that was me revising my patch in completely: this code used to read
```suggestion
coefficient_ring(A::MatAlgebra{T, S}) where {T, S} = A.coefficient_ring::base_ring_type(S)
```
which however requires a newer AA release adding that `base_ring_type`. So I meant to rewrite it, and apparently didn't finish (I was working on this stuff on an off for over a week, always half an hour here or there, always being interrupted, and last night I just decided to just submit it, no matter what. There is plenty more to be done, but I am not sure when I'll get to it)

File: src/QuadForm/Misc.jl [hunk 2]
```diff
@@ -879,7 +879,7 @@ function _find_quaternion_algebra(b, P, I)
     end
     M = vcat(M, v)
     push!(elts, f(L[i])) # cache
-    fl, w = can_solve_with_solution(M, target, side = :left)
+    fl, _ = can_solve_with_solution(M, target, side = :left)
     if fl
```
Comment (lgoettgens) [inline src/QuadForm/Misc.jl:883]:
```suggestion
    if can_solve(M, target, side = :left)
```
no need to compute the solution and then discard it

File: src/AlgAss/AlgMat.jl (hunk 1 above)
Comment (lgoettgens) [reply src/AlgAss/AlgMat.jl:15]: I prepared a new AA release in https://github.com/Nemocas/AbstractAlgebra.jl/pull/1708 to make the fix available

File: src/QuadForm/Misc.jl [hunk 3]
```diff
@@ -862,9 +862,9 @@ function _find_quaternion_algebra(b, P, I)
 
   found = false
   U, h = unit_group(R)
-  sign_vector = g -> begin
+  function sign_vector(g)
```
Comment (thofma) [inline src/QuadForm/Misc.jl:865]: why not anonymous?

Comment (lgoettgens) [reply src/QuadForm/Misc.jl:865]: My guess is that anonymous functions don't get a recognizable name in Jet analysis, but no real idea

**Pull request merged**